libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h

```diff
@@ -0,0 +1,525 @@
+        __VA_ARGS__ \
+    } else
+
+#define _LIBCUDACXX_HANDLE_POINTER_SPACE(name, ...) \
+    NV_IF_ELSE_TARGET( \
+        NV_IS_DEVICE, ( \
+            _LIBCUDACXX_HANDLE_SPACE(__is_cluster_shared,__cluster, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(__isShared, __shared, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(__isGlobal, __global, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(__is_grid_constant, __grid_constant, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(__isConstant, __constant, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(/* __isLocal when not broken */ bool, __local, name, __VA_ARGS__) \
+            {} \
+        ), ( \
+            _LIBCUDACXX_HANDLE_SPACE(bool, __global, name, __VA_ARGS__) \
+            {} \
+        ) \
+    )
+
+template<typename _Tag, _CUDA_VSTD::size_t _Value>
+struct __down_convertible_constant {
+    template<_CUDA_VSTD::size_t _OtherValue, typename = _CUDA_VSTD::__enable_if_t<(_OtherValue < _Value)>>
+    _LIBCUDACXX_INLINE_VISIBILITY
+    constexpr operator __down_convertible_constant<_Tag, _OtherValue>() const {
+        return {};
+    }
+
+    static const constexpr _CUDA_VSTD::size_t value = _Value;
+};
+
+struct __alignment_tag {};
+template<_CUDA_VSTD::size_t _Alignment>
+using __alignment = __down_convertible_constant<__alignment_tag, _Alignment>;
+
+template<_CUDA_VSTD::size_t _GuaranteedAlignment,
+    _CUDA_VSTD::size_t _MaxInterestingAlignment,
+    _CUDA_VSTD::size_t _MinInterestingAlignment,
+    _CUDA_VSTD::size_t _Alignment,
+    typename = void>
+struct __invoke_if_applicable {
+    template<typename _Fn>
+    _LIBCUDACXX_INLINE_VISIBILITY static async_contract_fulfillment __invoke(_Fn && __f) {
+        return _CUDA_VSTD::forward<_Fn>(__f)(__alignment<_Alignment>());
+    }
+};
+
+template<_CUDA_VSTD::size_t _GuaranteedAlignment,
+    _CUDA_VSTD::size_t _MaxInterestingAlignment,
+    _CUDA_VSTD::size_t _MinInterestingAlignment,
+    _CUDA_VSTD::size_t _Alignment>
+struct __invoke_if_applicable<_GuaranteedAlignment,
+    _MaxInterestingAlignment,
+    _MinInterestingAlignment,
+    _Alignment,
+    _CUDA_VSTD::__enable_if_t<
+        _Alignment >= _MaxInterestingAlignment
+            || _Alignment < _MinInterestingAlignment
+            || _Alignment < _GuaranteedAlignment
+>> {
+    template<typename _Fn>
+    _LIBCUDACXX_INLINE_VISIBILITY static async_contract_fulfillment __invoke(_Fn && __f) {
+        _LIBCUDACXX_UNREACHABLE();
+    }
+};
+
+template<_CUDA_VSTD::size_t _MaxInterestingAlignment,
+    _CUDA_VSTD::size_t _MinInterestingAlignment,
+    _CUDA_VSTD::size_t _GuaranteedAlignment,
+    typename _Fn>
+_LIBCUDACXX_INLINE_VISIBILITY async_contract_fulfillment __dispatch_alignment_bit(_Fn && __f, _CUDA_VSTD::size_t __alignment_fsb) {
+    _CUDA_VSTD::size_t __alignment_v = 1ull << (__alignment_fsb - 1);
+
+    if (__alignment_v >= _MaxInterestingAlignment) [[likely]] {
```

Comment:
I am not sure about those attributes. I fear we will need to _UGLIFY them via a macro

Reply:
Yeah, I have rewritten it with `__builtin_expect` for now, I do believe we'll need a proper macro for this.